edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs

```diff
@@ -387,79 +392,113 @@ static Option<string> GetCommonNameFromSubject(string subject)
 
-                    if (isPkcs8)
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    // On Windows the certificate in 'result' gives an error when used with kestrel: "No credentials are available in the security"
+                    // This is a suggested workaround that seems working (https://github.com/dotnet/runtime/issues/45680)
+                    result = new X509Certificate2(result.Export(X509ContentType.Pkcs12));
+
+                    // On Windows the imported certificate sometimes fails to use the private key and kestrel fails accepting connections (this is not related
+                    // to the other problem above). Try to access the private key and catch the error early. If it fails, re-importing the certificate
+                    // solves the problem:
+                    try
                     {
-                        key.ImportPkcs8PrivateKey(decodedKey, out _);
+                        if (oidRsaEncryption.Value == keyAlgorithm)
+                        {
+                            _ = result.GetRSAPrivateKey();
+                        }
+                        else
+                        {
+                            _ = result.GetECDsaPrivateKey();
+                        }
                     }
-                    else
+                    catch
                     {
-                        key.ImportECPrivateKey(decodedKey, out _);
+                        needsRetry = true;
                     }
-
-                    result = certificate.CopyWithPrivateKey(key);
-
-                    logger?.LogDebug("ECC private key has been imported and assigned to certificate");
                 }
-            }
-            catch (Exception ex)
-            {
-                var errorMessage = "Cannot import private key";
-                logger?.LogError(ex, errorMessage);
-                throw new InvalidOperationException(errorMessage, ex);
-            }
 
-            if (result == null)
-            {
-                var errorMessage = $"Cannot use certificate, not supported key algorithm: ${keyAlgorithm}";
-                logger?.LogError(errorMessage);
-                throw new InvalidOperationException(errorMessage);
-            }
+                if (!needsRetry)
+                {
+                    return result;
+                }
 
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-            {
-                // On Windows the certificate in 'result' gives an error when used with kestrel: "No credentials are available in the security"
-                // This is a suggested workaround that seems working (https://github.com/dotnet/runtime/issues/45680)
-                result = new X509Certificate2(result.Export(X509ContentType.Pkcs12));
+                logger?.LogWarning("Error importing certificate, retrying");
+                Thread.Sleep(TimeSpan.FromSeconds(1)); // Slow down retries a bit
```

Comment:
Consider sleeping for a shorter duration initially and then gradually increasing (i.e. exponential backoff). This is really only necessary if the underlying failure may be due to multiple clients making calls at the same time and competing for limited resources (e.g. retrieving private keys from the HSM via the security daemon?).

Reply:
Normally it would make sense, I see a few reasons that it would not help here:
- This thing is rare and every time I saw it failing, it always worked the second time
- It is not like there is a remote service or similar we use and maybe a transient network error prevents us connecting or similar. In those cases an exponential back off makes sense. In our case we use an OS functionality that should work every single time. In fact we don't why it fails at all. It should not.
- There is no specific reason I added a delay. It just felt right waiting a bit (and it did not do things worse, so why not). Also, the one sec is just a number. Could be 100ms or 2 sec. A little history here is that the very first version did not have counter (it never gave up) and in that case a continous  failure would have filled the logs very fast without a delay. Since we have a counter now, the worst could happen is that there is 10 lines in a block saying that error importing.

With all of that, I think adding an exponential backup would complicate the code with no clear benefit.

Reply:
Ok, that's fine with me.